### DISCUSSION TITLE
Suggestions for improvements & new ideas

### DISCUSSION THREAD

**Discussion by masurceac** [coltorapps/builder#2] [Ideas]
I would like to bring the following ideas into discussion:

## De-contextualizing Entities & Attributes
Currently, entities and attributes use context to pass the builder store instance, which makes it impossible to nest different builders in structures like: A > B > A.

## Option to Remove an Entity When Using createEntityComponent
It would be useful to have a .delete() method available in the props passed to a component when using createEntityComponent.

## Entity Tree Utilities
I've implemented some utilities myself, but I believe it would be beneficial to have a dedicated package that provides methods for manipulating an entity tree. For example: insert(schema, nodeToAppendTo), remove(startNode, endNode?), replace(newSchema, startNode, endNode?), etc.

## Async Validation
The `validate` methods are now async, which makes it really painful to call these methods inside react hook functions, like `useMemo` or `useState`. What about giving sync alternatives?

**Comment by sandulat** [comment]
> De-contextualizing Entities & Attributes

It's available since [0.2.0](https://github.com/coltorapps/builder/releases/tag/v0.2.0), #3.

> Option to Remove an Entity When Using createEntityComponent

I'm still unsure whether to make `createEntityComponent` handle methods that function exclusively in either a builder or an interpreter context.  

On one hand, `setValue` is a method that works only in an interpreter context, and does nothing in a builder context. So perhaps there's nothing wrong with having a `delete` method that functions only in a builder context and does nothing in an interpreter context. However, I'm still uncertain and open to discussion.

> Entity Tree Utilities

This is a bit out of scope and better suited for the developer’s code rather than the library. But if there’s enough demand, I’m happy to consider adding support for it.

> Async Validation

You can call the validation methods inside the new React 19 `use` hook, wrap your component with `Suspense`, and display a fallback (or nothing) while it's being validated. However, I’d need more context on your use case.
